core/base/precision_dispatch.hpp

```diff
@@ -0,0 +1,153 @@
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_BASE_PRECISION_DISPATCH_HPP_
+#define GKO_CORE_BASE_PRECISION_DISPATCH_HPP_
+
+
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/base/temporary_conversion.hpp"
+
+
+namespace gko {
+
+
+template <typename ValueType>
+detail::temporary_conversion<matrix::Dense<ValueType>>
+make_temporary_conversion(LinOp *matrix)
+{
+    auto result =
+        detail::temporary_conversion<matrix::Dense<ValueType>>::template create<
+            matrix::Dense<next_precision<ValueType>>>(matrix);
+    if (!result) {
+        GKO_NOT_SUPPORTED(matrix);
+    }
+    return result;
+}
+
+
+template <typename ValueType>
+detail::temporary_conversion<const matrix::Dense<ValueType>>
+make_temporary_conversion(const LinOp *matrix)
+{
+    auto result = detail::temporary_conversion<const matrix::Dense<ValueType>>::
+        template create<matrix::Dense<next_precision<ValueType>>>(matrix);
+    if (!result) {
+        GKO_NOT_SUPPORTED(matrix);
+    }
+    return result;
+}
+
+
+template <typename ValueType, typename Function, typename... Args>
+void precision_dispatch(Function fn, Args *... linops)
+{
+    fn(make_temporary_conversion<ValueType>(linops).get()...);
+}
+
+template <typename ValueType, typename Function>
+void precision_dispatch_spmv(Function fn, const LinOp *in, LinOp *out)
+{
+    // do we need to convert complex Dense to real Dense?
+    auto complex_to_real =
+        !(is_complex<ValueType>() ||
+          dynamic_cast<const ConvertibleTo<matrix::Dense<>> *>(in));
```

Comment:
In general I think this part really needs some explanation because it's a real trick. Either that or the longer implementation which checks the different combinations would be more clear. This reads (after integrating the not):
targeted type isn't complex and `in` isn't convertible to `Dense<double>`?

Basically, what I understand you are doing is depending on the fact that there is conversion to `Dense<next_precision<ValueType>>` which when float always evaluates to double, and vice versa. Thus, with the current implementation a `Dense` of `std::complex<double>` can convert to either `std::complex<double>` or  `std::complex<float>` and same for the simple `double` and `float` version. Thus, checking what you do here means checking against either `float` and `double` at the same time actually, although it appears to only check `double` at first.

This works of course but is maybe fragile (depends on the evolution of our conversion management) and is quite obfuscated. Something needs to be done to make this clearer at least a `WARN/NOTE` tag I think.